[WIP] Build docs (#2517)
DFHack/dfhack: docs/dev/building/Compile.rst

  /usr/local/Cellar/gcc@6/6.3.0/lib/gcc/6/i386/libstdc++.6.dylib  # for 32-bit DFHack

**Note:** If you build with a version of GCC that requires this, your DFHack
build will *not* be redistributable. (Even if you copy the ``libstdc++.6.dylib``
from your GCC version and distribute that too, it will fail on older OS X
versions.) For this reason, if you plan on distributing DFHack, it is highly
recommended to use GCC 4.8 or 7.

.. _osx-m1-notes:

Notes for M1 users
------------------

Alongside the above, you will need to follow these additional steps to get it
running on Apple silicon.

Install an x86 copy of ``homebrew`` alongside your existing one. `This
stackoverflow answer <https://stackoverflow.com/a/64951025>`__ describes the
process.

Follow the normal macOS steps to install ``cmake`` and ``gcc`` via your x86 copy of
``homebrew``. Note that this will install a GCC version newer than 7, so see
`osx-new-gcc-notes`.

In your terminal, ensure you have your path set to the correct homebrew in
addition to the normal ``CC`` and ``CXX`` flags above::

  export PATH=/usr/local/bin:$PATH

Docker
======

.. highlight:: bash

> You can use docker to build DFHack for windows. These instructions were developed
> on a linux host system.

Reviewer: nit: caps

```suggestion
You can use docker to build DFHack for Windows. These instructions were developed
on a Linux host system.
```

.. contents::
  :local:
  :depth: 1

Step 1: prepare a docker image
------------------------------

On your Linux host, install and run the docker daemon and then run these commands::

    xhost +local:root
    git clone https://github.com/BenLubar/build-env.git
    cd build-env/msvc
    docker build .
    docker image ls
    IMAGE_ID=<your image id>
    docker run -it --env="DISPLAY" --env="QT_X11_NO_MITSHM=1" --volume=/tmp/.X11-unix:/tmp/.X11-unix --user buildmaster --name dfhack-win $IMAGE_ID

The ``xhost`` command and ``--env`` parameters are there so you can eventually
run Dwarf Fortress from the container and have it display on your host.

Step 2: build DFHack
--------------------

> The ``docker run`` command above will give you a shell prompt (as the `buildmaster` user) in the

Reviewer: needs double backticks for code

```
/home/runner/work/dfhack/dfhack/docs/dev/building/Compile.rst:347: WARNING: undefined label: 'buildmaster'
```

  * ``Run 'git submodule update --init'`` when running CMake

  Submodules are a particularly confusing feature of Git. The
  `Git Book <https://git-scm.com/book/en/v2/Git-Tools-Submodules>`_ has a
  thorough explanation of them (as well as of many other aspects of Git) and
  is a recommended resource if you run into any issues. Other DFHack developers
  are also able to help with any submodule-related (or Git-related) issues
  you may encounter.

Dependencies
------------

If you haven't already checked, this would be a good point to ensure you have
all the `dependencies <build-dependencies>`.

Contributing to DFHack
----------------------

For details on contributing to DFHack, including pull requests, code
format, and more, please see `contributing-code`.

Build settings
==============

This section describes build configuration options that apply to all platforms.
If you don't have a working build environment set up yet, follow the instructions
in the platform-specific sections below first, then come back here.

Be sure to check out common `build options <build-options>`.

Instructions
============

.. contents::
  :local:
  :depth: 1

> 

Reviewer 2: since this is the main focus of the page now, I think we can get rid of this header and move everything below it up a header level.

Author: I disagree, this makes for a good landing spot for anyone looking for just the instructions. The information would of course be visible from the top level, but it does add clutter. Have you taken a look at the html? The index links to such sections.

Reviewer 2: I did most of the reviewing from the generated HTML. The current TOC on the Building page looks like this:
```
Compilation
How to get the code
Build settings
Instructions
```

I'm saying this would save readers a click if instead of "Instructions" we let them directly access the instructions for the platform they're interested in.

Author: You've convinced me actually. I had considered this originally, but decided against it because I wanted to reinforce that there are sections relevant to everyone. I thought all the submodules being a gotcha was a good reason to avoid adding platforms for users to instantly spot and click and end up not populating submodules. So that was my rational, if you agree with it I'll stick with it otherwise move them up a level

Reviewer 2: yeah, I see what you're saying, and that's a good point. How about this:
- Keep `How to get the code` at the top
- Remove the `Dependencies` and `Build Settings` sections (which are both just stubs pointing to other pages)
- Move the platforms up a level (get rid of the `Instructions` header)
- At the start of each platform section, have a short sentence like:
```
Be sure you have `downloaded the code <How to get the code>`_, installed the `build-dependencies`, and configured the `build-options`.
```
Then even if they click straight to their platform, they won't miss common information.

Author: That would take us back to that idea of "including the same text in multiple locations"

Reviewer 2: yeah, but this "repeated text" is just hyperlinks to other sections, not actual original content. I don't think that crosses the threshold for "duplication"

.. _compile-linux:

Linux
-----
On Linux, DFHack acts as a library that shadows parts of the SDL API using LD_PRELOAD.

Build
~~~~~
Building is fairly straightforward. Enter the ``build`` folder (or create an
empty folder in the DFHack directory to use instead) and start the build like this::

    cd build
    cmake .. -G Ninja -DCMAKE_BUILD_TYPE:string=Release -DCMAKE_INSTALL_PREFIX=<path to DF>
    ninja install  # or ninja -jX install to specify the number of cores (X) to use

<path to DF> should be a path to a copy of Dwarf Fortress, of the appropriate
version for the DFHack you are building. This will build the library along
with the normal set of plugins and install them into your DF folder.

Alternatively, you can use ccmake instead of cmake::

    cd build
    ccmake .. -G Ninja
    ninja install

This will show a curses-based interface that lets you set all of the
extra options. You can also use a cmake-friendly IDE like KDevelop 4
or the cmake-gui program.

> Windows cross compiling from Linux

Reviewer 2: I think this should be on the same header level as the main operating systems, since it is a significantly different use case

Author: How about under the heading: Docker?
Cause otherwise I have the differing opinion that this is the best hierarchy.

Reviewer 2: I think I'd be happy as long as it shows up in the TOC on this page.

Reviewer: My vote is for "Windows cross-compiling from Linux". No opinion on the hierarchy (nested or sibling with "Windows" are fine with me).

Author: I need a tie breaker

Reviewer: Ok, my vote is "if it ain't broke, don't fix it". So `Docker` => `Windows cross compiling from Linux`. The instructions only apply to building for Windows currently, so `Docker` is ambiguous.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. highlight:: bash

If you are on Linux but need to produce a Windows build (for example, because the
DF version you're working on isn't out for Linux yet), here is how you can build
and run a Windows binary on Linux.

.. contents::
  :local:
  :depth: 1

Step 1: prepare a docker image
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

On your Linux host, install and run the docker daemon and then run these commands::

    xhost +local:root
    git clone https://github.com/BenLubar/build-env.git
    cd build-env/msvc
    docker build .
    docker image ls
    IMAGE_ID=<your image id>
    docker run -it --env="DISPLAY" --env="QT_X11_NO_MITSHM=1" --volume=/tmp/.X11-unix:/tmp/.X11-unix --user buildmaster --name dfhack-win $IMAGE_ID

The ``xhost`` command and ``--env`` parameters are there so you can eventually
run Dwarf Fortress from the container and have it display on your host.

Step 2: build DFHack
^^^^^^^^^^^^^^^^^^^^

> The ``docker run`` command above will give you a shell prompt (as root) in the

Reviewer 2: ```suggestion
The ``docker run`` command above will give you a shell prompt (as the `buildmaster` user) in the
```

container. Inside the container, run the following commands::

    git clone https://github.com/DFHack/dfhack.git
    cd dfhack
    git submodule update --init
    cd build
    dfhack-configure windows 64 Release
    dfhack-make

Inside the ``dfhack-*`` scripts there are several commands that set up the wine
server. Each invocation of a windows tool will cause wine to run in the container.
Preloading the wineserver and telling it not to exit will speed configuration and
compilation up considerably (approx. 10x). You can configure and build DFHack
with regular ``cmake`` and ``ninja`` commands, but your build will go much slower.

> Step 3: copy Dwarf Fortress to the container

Reviewer 2: ```suggestion
Step 3: copy or mount Dwarf Fortress to the container
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

First, create a directory in the container to house the Dwarf Fortress binary and
assets::

    mkdir ~/df

> If you can just downlaod Dwarf Fortress directly into the container, then that's fine.

Reviewer 2: ```suggestion
If you can just download Dwarf Fortress directly into the container, then that's fine.
```

Otherwise, you can do something like this in your host Linux environment to copy an
installed version to the container::

    cd ~/.steam/steam/steamapps/common/Dwarf\ Fortress/
    docker cp . dfhack-win:df/

Step 4: install DFHack and run DF
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Back in the container, run the following commands::

    cd dfhack/build
    cmake .. -DCMAKE_INSTALL_PREFIX=/home/buildmaster/df
    ninja install
    cd ~/df
    wine64 "Dwarf Fortress.exe"

Other notes
^^^^^^^^^^^

Closing your shell will kick you out of the container. Run this command on your Linux
host when you want to reattach::

    docker start -ai dfhack-win

> If you edit code and need to rebuild, run ``dfhack-make`` and then ``ninja install``.

Reviewer 2: ```suggestion
If you edit code and need to rebuild, run ``dfhack-make install``.
```
